Modules/_abc.c

```diff
@@ -643,31 +642,20 @@ _abc__abc_subclasscheck_impl(PyObject *module, PyObject *self,
     }
     Py_DECREF(ok);
 
-    /* 4. Check if it's a direct subclass.
-     *
-     * if cls in getattr(subclass, '__mro__', ()):
-     *     cls._abc_cache.add(subclass)
-     *     return True
-     */
-    if (_PyObject_LookupAttrId(subclass, &PyId___mro__, &mro) < 0) {
-        goto end;
-    }
-    if (mro != NULL) {
-        if (!PyTuple_Check(mro)) {
-            // Python version supports non-tuple iterable.  Keep it as
-            // implementation detail.
-            PyErr_SetString(PyExc_TypeError, "__mro__ is not a tuple");
+    /* 4. Check if it's a direct subclass. */
+    PyObject *mro = ((PyTypeObject *)subclass)->tp_mro;
+    assert(PyTuple_Check(mro));
+    for (pos = 0; pos < PyTuple_GET_SIZE(mro); pos++) {
+        PyObject *mro_item = PyTuple_GET_ITEM(mro, pos);
+        if (mro_item == NULL) {
```

Comment:
Late review, this check is redundant or not right. To keep it, you need to set an exception in the block, otherwise you are returning NULL without an exception, then SystemError.